src/analyses/wrapperFunctionAnalysis.ml

```diff
@@ -0,0 +1,257 @@
+        else p + 1 (* Unique addresses + top address *)
+
+      let names x = if x = (n () - 1) then "top" else Format.asprintf "%d" x
+
+    end)
+
+  (* Map for counting function call node visits up to n (of the current thread). *)
+  module UniqueCallCounter = struct
+    include MapDomain.MapBot_LiftTop(Q.NodeFlatLattice)(Chain)
+
+    (* Increase counter for given node. If it does not exist yet, create it. *)
+    let add_unique_call counter node =
+      let unique_call = `Lifted node in
+      let count = find unique_call counter in
+      let c' = if Chain.is_top count then
+        counter
+      else
+        remove unique_call counter |> add unique_call (count + 1)
+      in
+      if dbg then dpf"add_unique_call node=<%a> count_before=%d count_after=%d" Node.pretty node count (find unique_call c');
+      c'
+  end
+
+  module D = Lattice.Prod (UniqueCallCounter) (Q.NodeFlatLattice)
+  module C = D
+
+  let wrappers = Hashtbl.create 13
+
+  (* transfer functions *)
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    ctx.local
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    ctx.local
+
+  let body ctx (f:fundec) : D.t =
+    ctx.local
+
+  let return ctx (exp:exp option) (f:fundec) : D.t =
+    if dbg then dpf"return f=<%a>" CilType.Fundec.pretty f;
+    ctx.local
+
+  let enter ctx (lval: lval option) (f:fundec) (args:exp list) : (D.t * D.t) list =
+    st "enter" ctx;
+    let counter, wrapper_node = ctx.local in
+    let new_wrapper_node =
+      if Hashtbl.mem wrappers f.svar.vname then
+        begin
+        if dbg then dpf"is wrapper";
+        match wrapper_node with
+        (* if an interesting callee is called by an interesting caller, then we remember the caller context *)
+        | `Lifted _ -> wrapper_node
+          (* if dbg then dpf"interesting caller, keep caller context"; *)
+        (* if an interesting callee is called by an uninteresting caller, then we remember the callee context *)
+        | _         -> `Lifted (node_for_ctx ctx)
+          (* if dbg then dpf"uninteresting caller, keep callee context"; *)
+        end
+      else
+        Q.NodeFlatLattice.top () (* if an uninteresting callee is called, then we forget what was called before *)
+    in
+    let callee = (counter, new_wrapper_node) in
+    [(ctx.local, callee)]
+
+  let combine_env ctx lval fexp f args fc (counter, _) f_ask =
+    if dbg then dpf"combine f=<%a>" CilType.Fundec.pretty f;
+    (* Keep (potentially higher) counter from callee and keep wrapper node from caller *)
+    let _, lnode = ctx.local in
+    (counter, lnode)
+
+  let combine_assign ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc ((counter, _):D.t) (f_ask: Queries.ask) : D.t =
+    ctx.local
+
+  let add_unique_call ctx =
+    let counter, wrapper_node = ctx.local in
+    (* TODO: previously, unique count isn't by wrapper node (e.g. my_malloc_wrapper), but by wrapped node (e.g. malloc). Why, and is it safe to change? *)
```

Comment:
This is probably incidental and due to the student who made the implementation of the unique counter. Should be safe to change!